--- modules/contrib/imce/src/ImceFileField.php	2015-07-31 09:17:09.000000000 +0200
+++ modules/contrib/imce/src/ImceFileField.php	2015-07-31 09:17:43.000000000 +0200
@@ -67,6 +67,7 @@
    * Processes widget form.
    */
   public static function processWidget($element, FormStateInterface $form_state, $form) {
+    $element['#id'] = $element['#attributes']['data-drupal-selector'];
     // Button
     $element['imce_button'] = array(
       '#name' => $element['#name'] . '[imce_button]',
